pkg/controllers/vdb/obj_reconciler.go

```diff
@@ -468,49 +483,113 @@ func (o *ObjReconciler) checkVProxyConfigMap(ctx context.Context, cmName types.N
-		// Create or update the client proxy deployment
-		vpErr := o.checkVProxyDeployment(ctx, sc)
-		if vpErr != nil {
-			return ctrl.Result{}, vpErr
+// updateVProxyDeployment will update a vproxy deployment if any field
+// has changed since the last time.
+func (o *ObjReconciler) updateVProxyDeployment(ctx context.Context, sts *appsv1.StatefulSet,
+	curDep *appsv1.Deployment, sc *vapi.Subcluster) error {
+	if !vmeta.UseVProxy(o.Vdb.Annotations) {
+		return nil
+	}
+	vpName := names.GenVProxyName(o.Vdb, sc.Name)
+	expDep := builder.BuildVProxyDeployment(vpName, o.Vdb, sc)
+	if *sts.Spec.Replicas == 0 {
+		*expDep.Spec.Replicas = 0
+	}
+	return o.updateDep(ctx, curDep, expDep)
+}
+
+// ensureStsExists checks if the sts exists and will create it if it does not.
+func (o *ObjReconciler) ensureStsExists(ctx context.Context, nm types.NamespacedName, curSts, expSts *appsv1.StatefulSet) (bool, error) {
+	err := o.Rec.GetClient().Get(ctx, nm, curSts)
+	if err != nil && kerrors.IsNotFound(err) {
+		o.Log.Info("Creating statefulset", "Name", nm, "Size", expSts.Spec.Replicas, "Image", expSts.Spec.Template.Spec.Containers[0].Image)
+		o.PFacts.Invalidate()
+		if stsErr := createSts(ctx, o.Rec, expSts, o.Vdb); stsErr != nil {
+			return false, stsErr
 		}
+		return true, nil
+	}
+	return false, nil
+}
+
+// reconcileVproxy will check if the deployment and its configmap exist and create
+// them.
+func (o *ObjReconciler) reconcileVProxy(ctx context.Context, sc *vapi.Subcluster) (*appsv1.Deployment, error) {
+	if err := o.checkVProxyConfigMap(ctx, sc); err != nil {
+		return nil, err
 	}
+	return o.createVProxyDeployment(ctx, sc)
+}
 
+// deleteVProxy deletes the proxy deployment and configmap ig they exist
+func (o *ObjReconciler) deleteVProxy(ctx context.Context, scName string) error {
+	err := o.deleteVProxyDeployment(ctx, scName)
+	if err != nil {
+		return err
+	}
+	return o.deleteVProxyConfigMapIfExists(ctx, scName)
+}
+
+// reconcileSts reconciles the statefulset for a particular subcluster.  Returns
+// true if any create/update was done.
+func (o *ObjReconciler) reconcileSts(ctx context.Context, sc *vapi.Subcluster) (ctrl.Result, error) {
 	// Create or update the statefulset
 	nm := names.GenStsName(o.Vdb, sc)
 	curSts := &appsv1.StatefulSet{}
 	expSts := builder.BuildStsSpec(nm, o.Vdb, sc)
-	err := o.Rec.GetClient().Get(ctx, nm, curSts)
-	if err != nil && kerrors.IsNotFound(err) {
-		o.Log.Info("Creating statefulset", "Name", nm, "Size", expSts.Spec.Replicas, "Image", expSts.Spec.Template.Spec.Containers[0].Image)
-		// Invalidate the pod facts cache since we are creating a new sts
-		o.PFacts.Invalidate()
-		return ctrl.Result{}, createSts(ctx, o.Rec, expSts, o.Vdb)
+	stsCreated, err := o.ensureStsExists(ctx, nm, curSts, expSts)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	var curDep *appsv1.Deployment
+	if vmeta.UseVProxy(o.Vdb.Annotations) {
+		if curDep, err = o.reconcileVProxy(ctx, sc); err != nil {
```

Comment:
We should only create proxy deployment after a subcluster sts is created. Do reconcileVProxy() only in line 591 and line 648.

Reply:
I don't need to. reconcileVProxy will create the proxy deployment whenever it is missing. If you look at the code, you will see I always check first if the proxy exists
- If I move this to line 591, if I manually delete the deployment, it won't get recreated because the sts already exists.
- In line 648, I only want to recreate the sts and that has nothing to do with the proxy deployment. Keep in mind sts no longer owns the deployment so deleting it won't delete the proxy.

Reply:
I don't want the proxy to be created when a statefulset is updated. During online upgrade, we removed the old main and the proxy. However, when we rename subclusters in the new main, the old proxy will be created again due to a statefulset update. 

We probably need a more sophisticated proxy reconciler to handle the rename case.

Reply:
Or we can reuse the annotation trick. I think what you want is that when we rename the subclusters, the stss and proxies do not get rebuilt/recreated.